pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaManaged.hpp

```diff
@@ -1419,6 +2125,153 @@ class StateVectorCudaManaged
+         }},
+        {"SingleExcitationPlus",
+         [&](auto &&controlled_wires, auto &&controlled_values, auto &&wires,
+             auto &&adjoint) {
+             return applyControlledGeneratorSingleExcitationPlus(
+                 std::forward<decltype(controlled_wires)>(controlled_wires),
+                 std::forward<decltype(controlled_values)>(controlled_values),
+                 std::forward<decltype(wires)>(wires),
+                 std::forward<decltype(adjoint)>(adjoint));
+         }},
+        {"DoubleExcitation",
+         [&](auto &&controlled_wires, auto &&controlled_values, auto &&wires,
+             auto &&adjoint) {
+             return applyControlledGeneratorDoubleExcitation(
+                 std::forward<decltype(controlled_wires)>(controlled_wires),
+                 std::forward<decltype(controlled_values)>(controlled_values),
+                 std::forward<decltype(wires)>(wires),
+                 std::forward<decltype(adjoint)>(adjoint));
+         }},
+        {"DoubleExcitationMinus",
+         [&](auto &&controlled_wires, auto &&controlled_values, auto &&wires,
+             auto &&adjoint) {
+             return applyControlledGeneratorDoubleExcitationMinus(
+                 std::forward<decltype(controlled_wires)>(controlled_wires),
+                 std::forward<decltype(controlled_values)>(controlled_values),
+                 std::forward<decltype(wires)>(wires),
+                 std::forward<decltype(adjoint)>(adjoint));
+         }},
+        {"DoubleExcitationPlus",
+         [&](auto &&controlled_wires, auto &&controlled_values, auto &&wires,
+             auto &&adjoint) {
+             return applyControlledGeneratorDoubleExcitationPlus(
+                 std::forward<decltype(controlled_wires)>(controlled_wires),
+                 std::forward<decltype(controlled_values)>(controlled_values),
+                 std::forward<decltype(wires)>(wires),
+                 std::forward<decltype(adjoint)>(adjoint));
+         }},
+        {"GlobalPhase",
+         [&](auto &&controlled_wires, auto &&controlled_values, auto &&wires,
+             auto &&adjoint) {
+             return applyControlledGeneratorGlobalPhase(
+                 std::forward<decltype(controlled_wires)>(controlled_wires),
+                 std::forward<decltype(controlled_values)>(controlled_values),
+                 std::forward<decltype(wires)>(wires),
+                 std::forward<decltype(adjoint)>(adjoint));
+         }},
+        {"MultiRZ", [&](auto &&controlled_wires, auto &&controlled_values,
+                        auto &&wires, auto &&adjoint) {
+             return applyControlledGeneratorMultiRZ(
+                 std::forward<decltype(controlled_wires)>(controlled_wires),
+                 std::forward<decltype(controlled_values)>(controlled_values),
+                 std::forward<decltype(wires)>(wires),
+                 std::forward<decltype(adjoint)>(adjoint));
+         }}};
+
```

Comment:
I see a lot of repetition here. Do you think we can do something like below?
```suggestion
    // Generator forwarding logic.
    template <typename Func>
    auto makeControlledGenerator(Func&& func) {
        return [func = std::forward<Func>(func)](auto &&controlled_wires,
                                                 auto &&controlled_values,
                                                 auto &&wires,
                                                 auto &&adjoint) {
            return func(std::forward<decltype(controlled_wires)>(controlled_wires),
                        std::forward<decltype(controlled_values)>(controlled_values),
                        std::forward<decltype(wires)>(wires),
                        std::forward<decltype(adjoint)>(adjoint));
        };
    }
    // Holds the mapping from controlled-gate labels to associated generator functions.
  const CGMap controlled_generator_map_{
      {"PhaseShift", makeControlledGenerator(applyControlledGeneratorPhaseShift)},
      {"RX", makeControlledGenerator(applyControlledGeneratorRX)},
      {"RY", makeControlledGenerator(applyControlledGeneratorRY)},
      {"RZ", makeControlledGenerator(applyControlledGeneratorRZ)},
      {"IsingXX", makeControlledGenerator(applyControlledGeneratorIsingXX)},
      {"IsingXY", makeControlledGenerator(applyControlledGeneratorIsingXY)},
      {"IsingYY", makeControlledGenerator(applyControlledGeneratorIsingYY)},
      {"IsingZZ", makeControlledGenerator(applyControlledGeneratorIsingZZ)},
      {"SingleExcitation", makeControlledGenerator(applyControlledGeneratorSingleExcitation)},
      {"SingleExcitationMinus", makeControlledGenerator(applyControlledGeneratorSingleExcitationMinus)},
      {"SingleExcitationPlus", makeControlledGenerator(applyControlledGeneratorSingleExcitationPlus)},
      {"DoubleExcitation", makeControlledGenerator(applyControlledGeneratorDoubleExcitation)},
      {"DoubleExcitationMinus", makeControlledGenerator(applyControlledGeneratorDoubleExcitationMinus)},
      {"DoubleExcitationPlus", makeControlledGenerator(applyControlledGeneratorDoubleExcitationPlus)},
      {"GlobalPhase", makeControlledGenerator(applyControlledGeneratorGlobalPhase)},
      {"MultiRZ", makeControlledGenerator(applyControlledGeneratorMultiRZ)}
  };
```

Reply:
Thanks for the suggestion Amintor! I have just tried to use this, but because e.g. `applyControlledGeneratorPhaseShift` is a non-static member function, I can't do `makeControlledGenerator(applyControlledGeneratorPhaseShift)` straightaway. I need to bind it to a lambda first to make it work:
```
const CGMap controlled_generator_map_{
        {"PhaseShift", makeControlledGenerator([this](const auto& c_wires,
                   const auto& c_values,
                   const auto& wires,
                   bool adj) -> PrecisionT{return applyControlledGeneratorPhaseShift(c_wires, c_values, wires, adj);})},
                   ...
```

This does work, do we think this is more or less neat than the original implementation? :thinking:

Reply:
Could you please provide more context here?

Reply:
Of course, sorry for the delay!

so if I apply the suggestion I will get an error when compiling:
```
StateVectorCudaManaged.hpp:2143:46: error: invalid use of non-static member function ‘Pennylane::LightningGPU::StateVectorCudaManaged<Precision>::PrecisionT Pennylane::LightningGPU::StateVectorCudaManaged<Precision>::applyControlledGeneratorPhaseShift(const std::vector<long unsigned int>&, const std::vector<bool>&, const std::vector<long unsigned int>&, bool) [with Precision = double; Pennylane::LightningGPU::StateVectorCudaManaged<Precision>::PrecisionT = double]’
 2143 |       {"PhaseShift", makeControlledGenerator(applyControlledGeneratorPhaseShift)},
      |                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/joseph/work/pennylane-lightning/pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaManaged.hpp:1706:23: note: declared here
 1706 |     inline PrecisionT applyControlledGeneratorPhaseShift(
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I have tried using std::mem_fn to fix this but I run into:
```
/home/joseph/work/pennylane-lightning/pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaManaged.hpp:2134:24: error: no match for call to ‘(const std::_Mem_fn<double (Pennylane::LightningGPU::StateVectorCudaManaged<double>::*)(const std::vector<long unsigned int>&, const std::vector<bool>&, const std::vector<long unsigned int>&, bool)>) (const std::vector<long unsigned int>&, const std::vector<bool>&, const std::vector<long unsigned int>&, bool)’
 2134 |             return func(std::forward<decltype(controlled_wires)>(controlled_wires),
      |                    ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 2135 |                         std::forward<decltype(controlled_values)>(controlled_values),
      |                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 2136 |                         std::forward<decltype(wires)>(wires),
      |                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 2137 |                         std::forward<decltype(adjoint)>(adjoint));
      |                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The way I got it to work was to bind it to a lambda, but maybe there's a simpler solution I'm not seeing...

Reply:
I have found how to make the `makeControlledGenerator` function accept member functions now:

```
// Generator forwarding logic.  
    template <typename Func>  
    auto makeControlledGenerator(Func&& func) {  
        return [this, func = std::forward<Func>(func)](auto &&controlled_wires,  
                                                 auto &&controlled_values,  
                                                 auto &&wires,  
                                                 auto &&adjoint) {  
            return (this->*func)(std::forward<decltype(controlled_wires)>(controlled_wires),  
                        std::forward<decltype(controlled_values)>(controlled_values),  
                        std::forward<decltype(wires)>(wires),  
                        std::forward<decltype(adjoint)>(adjoint));  
        };  
    }
    // Holds the mapping from controlled-gate labels to associated generator
    // functions.
    const CGMap controlled_generator_map_{
        {"PhaseShift", makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorPhaseShift)},
        {"RX",makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorRX)},
        {"RY",makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorRY)},
        {"RZ",makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorRZ)},
        {"IsingXX",makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorIsingXX)},
        {"IsingXY",makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorIsingXY)},
        {"IsingYY",makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorIsingYY)},
        {"IsingZZ",makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorIsingZZ)},
        {"SingleExcitation",makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorSingleExcitation)},
        {"SingleExcitationMinus",makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorSingleExcitationMinus)},
        {"SingleExcitationPlus",makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorSingleExcitationPlus)},
        {"DoubleExcitation",makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorDoubleExcitation)},
        {"DoubleExcitationMinus",makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorDoubleExcitationMinus)},
        {"DoubleExcitationPlus",makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorDoubleExcitationPlus)},
        {"GlobalPhase",makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorGlobalPhase)},
{"MultiRZ", makeControlledGenerator(&StateVectorCudaManaged::applyControlledGeneratorMultiRZ) }};
```

Reply:
This is great! Thank you for that!